odxtools/environmentdatadescription.py

```diff
@@ -177,40 +203,64 @@ def decode_from_pdu(self, decode_state: DecodeState) -> ParameterValue:
         """Extract the bytes from a PDU and convert them to a physical value.
         """
 
-        # retrieve the relevant DTC parameter which must be located in
-        # front of the environment data description.
+        # retrieve the DTC as a numerical value from the referenced
+        # parameter (which must be located somewhere before the
+        # parameter using the environment data description)
         if self.param_snref is None:
             odxraise("Specifying the DTC parameter for environment data "
                      "descriptions via SNPATHREF is not supported yet")
             return None
 
-        dtc_param: Optional[Parameter] = None
-        dtc_dop: Optional[DtcDop] = None
-        dtc_param_value: Optional[ParameterValue] = None
+        numerical_dtc_value: Optional[ParameterValue] = None
         for prev_param, prev_param_value in reversed(decode_state.journal):
             if prev_param.short_name == self.param_snref:
-                dtc_param = prev_param
-                prev_dop = getattr(prev_param, "dop", None)
-                if not isinstance(prev_dop, DtcDop):
-                    odxraise(f"The DOP of the parameter referenced by environment data "
-                             f"descriptions must be a DTC-DOP (is '{type(prev_dop).__name__}')")
+                if not isinstance(prev_param, ValueParameter):
+                    odxraise(
+                        f"The parameter referenced by environment data descriptions "
+                        f"must use a VALUE-PARAMETER (encountered {type(prev_param).__name__} "
+                        f"for reference '{self.param_snref}' of ENV-DATA-DESC '{self.short_name}')")
                     return
-                dtc_dop = prev_dop
-                dtc_param_value = prev_param_value
+
+                prev_dop = prev_param.dop
+                if isinstance(prev_dop, DtcDop):
+                    if prev_dop.diag_coded_type.base_data_type != DataType.A_UINT32:
+                        odxraise(f"The data type used by the DOP of the parameter referenced "
+                                 f"by environment data descriptions must be A_UINT32 "
+                                 f"(encountered '{prev_dop.diag_coded_type.base_data_type.value}')")
+                        return
+
+                    if prev_param_value is None:
+                        odxraise()
+                        return
+
+                    numerical_dtc_value = prev_dop.convert_to_numerical_trouble_code(
```

Comment:
this block feels copy-pasted

Reply:
Introduce a function, that takes `prev_dop` and `prev_param_value` as input and return `numerical_dtc_value`

Reply:
good point. Done: https://github.com/mercedes-benz/odxtools/pull/379/commits/a9a791f804540f070914559fb391bb0dca4dda4e